KT-45777: Implement cache eviction for classpath snapshot cache

hungvietnguyen opened the pull request.

gavra0 commented on compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/InMemoryCacheWithEviction.kt:
```diff
@@ -0,0 +1,133 @@
+     * Contains recently-used cache entries. We keep strong references to the values so that they can always be used until they are moved to
+     * [softCache], where they can be used for some more time until being garbage collected.
+     */
+    @GuardedBy("lock")
+    private val strongCache = mutableMapOf<KEY, VALUE>()
+
+    /** Contains older cache entries. We keep [SoftReference]s to the values so that they can be used until being garbage collected. */
+    @GuardedBy("lock")
+    private val softCache = mutableMapOf<KEY, SoftReference<VALUE>>()
+
+    /** Maps each key in [strongCache] to the most recent time period when it is used. */
+    @GuardedBy("lock")
+    private val cacheEntriesLastUsed = mutableMapOf<KEY, Int>()
```
WDYT about having a single ConcurrentHashMap to contain KEY -> VALUE entries, and each entry is just a wrapper that has:
- a strong reference to cached value
- a weak ref to cached value
- last used timestamp

it may simplify things.

Also, can you please add tests for this cache (you probably need to make memory usage injectable).

hungvietnguyen replied:
That's a great idea! It does simplify this class greatly (even though there is still heavy use of synchronization that I can't manage to avoid).

One tiny issue is that cache eviction will be slightly slower: with the previous approach, we just looked at the strong references, whereas with the new approach, we're going through all entries including both strong and soft references.

I added a unit test too.

gavra0 reviewed (commented):
Overall, it looks good. With the eviction policy in this PR, what is the hit rate for e.g jetbrains/kotlin build or any other project?

hungvietnguyen commented:
> Overall, it looks good. With the eviction policy in this PR, what is the hit rate for e.g jetbrains/kotlin build or any other project?

I tested it on the usual ABI change scenario in `stdlib` of `jetbrains/kotlin` and got 25 cache misses [1] / 7944 cache queries. Loading time is reduced from 3.5s to 0.93s [2].

[1] 25 cache misses because there are 25 not up-to-date transforms (including the changed subproject and `kotlin-compiler-embeddable.jar`, and some smaller ones for some reason -- I've added a note to investigate this later but it's orthogonal to this feature).

[2] 0.93s includes:
   - Loading 25 classpath snapshot entries due to cache misses: 0.35s
   - Loading cache hits (~8000) (including waiting on locks) + combining classpath entries into classpath snapshot: 0.48 s
   - Cache eviction: 0.04 s
   - Misc work (e.g., collecting metrics): 0.06s
   
 =============
 
 I also just added a small cleanup commit.

hungvietnguyen commented:
Squashed cleanup commit with the main commit so it's easier for JB to review.

Tapchicoma commented:
Related issue: https://youtrack.jetbrains.com/issue/KT-51978

hungvietnguyen commented:
Rebased and split the cleanup commit into 2 smaller commits.

Tapchicoma commented:
Merged into `master` as: 6d3e679a5982db914785f4e056db04d0dff1856f, c109d3d95bd140eba609539203733155beba06ad and 862c0b0c9a5c3335ce08cc1d1c9c934cd09aa98e commits.
